File: PMEditor/src/nl/adaptivity/diagram/android/LWDrawableView.java
package nl.adaptivity.diagram.android;

import nl.adaptivity.diagram.Drawable;
import nl.adaptivity.diagram.DrawingStrategy;
import nl.adaptivity.diagram.Pen;
import nl.adaptivity.diagram.Rectangle;
import nl.adaptivity.diagram.Theme;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.RectF;


public class LWDrawableView implements LightView{

  private Drawable aItem;
  private AndroidCanvas aAndroidCanvas;

  public LWDrawableView(Drawable pItem) {
    aItem = pItem;
  }

  @Override
  public void setFocussed(boolean pFocussed) {
    if (pFocussed) {
      aItem.setState(aItem.getState()|Drawable.STATE_FOCUSSED);
    } else {
      aItem.setState(aItem.getState()& ~Drawable.STATE_FOCUSSED);
    }
  }

  @Override
  public boolean getFocussed() {
    return (aItem.getState()&Drawable.STATE_FOCUSSED)!=0;
  }

  @Override
  public void setSelected(boolean pSelected) {
    if (pSelected) {
      aItem.setState(aItem.getState()|Drawable.STATE_SELECTED);
    } else {
      aItem.setState(aItem.getState()& ~Drawable.STATE_SELECTED);
    }
  }

  @Override
  public boolean getSelected() {
    return (aItem.getState()&Drawable.STATE_SELECTED)!=0;
  }

  @Override
  public void setTouched(boolean pTouched) {
    if (pTouched) {
      aItem.setState(aItem.getState()|Drawable.STATE_TOUCHED);
    } else {
      aItem.setState(aItem.getState()& ~Drawable.STATE_TOUCHED);
    }
  }

  @Override
  public boolean getTouched() {
    return (aItem.getState()&Drawable.STATE_TOUCHED)!=0;
  }

  @Override
  public void getBounds(RectF pRect) {
    Rectangle bounds = aItem.getBounds();
    pRect.set(bounds.leftf(), bounds.topf(), bounds.rightf(), bounds.bottomf());
  }


  /**
   * Craw this drawable onto an android canvas. The canvas has an ofset
   * preapplied so the top left of the drawing is 0,0.
   */
  @Override
  public <S extends DrawingStrategy<S, AndroidPen, AndroidPath>> void draw(Canvas pCanvas, Theme<S, AndroidPen, AndroidPath> pTheme, double pScale) {
    if (aAndroidCanvas==null) {
      aAndroidCanvas=new AndroidCanvas(pCanvas, pTheme);
    } else {
      aAndroidCanvas.setCanvas(pCanvas);
    }
//    if ((aItem.getState()&Drawable.STATE_SELECTED)!=0) {
//      Rectangle bounds = aItem.getBounds();
//      Bitmap bitmap = Bitmap.createBitmap((int) Math.ceil(bounds.width*pScale), (int) Math.ceil(bounds.height*pScale), Bitmap.Config.ARGB_8888);
//      AndroidCanvas bitmapCanvas = new AndroidCanvas(new Canvas(bitmap), pTheme);
//      aItem.draw(bitmapCanvas.scale(pScale), null);
//      AndroidPen pen = pTheme.getPen(AndroidExtraThemeItem.BLUR, aItem.getState());
//      aAndroidCanvas.drawBitmap(0f, 0f, bitmap, pen);
//    } else {
    aItem.draw(aAndroidCanvas.scale(pScale), null);
//    }
  }

}

File: PMEditor/src/main/java/nl/adaptivity/diagram/android/LWDrawableView.kt
/*
 * Copyright (c) 2017.
 *
 * This file is part of ProcessManager.
 *
 * ProcessManager is free software: you can redistribute it and/or modify it under the terms of version 3 of the
 * GNU Lesser General Public License as published by the Free Software Foundation.
 *
 * ProcessManager is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License along with ProcessManager.  If not,
 * see <http://www.gnu.org/licenses/>.
 */

package nl.adaptivity.diagram.android

import android.graphics.Canvas
import android.graphics.RectF
import nl.adaptivity.diagram.Drawable
import nl.adaptivity.diagram.Rectangle
import nl.adaptivity.diagram.Theme


open class LWDrawableView(open val item: Drawable) : LightView {
    /** Cached canvas  */
    private var androidCanvas: AndroidCanvas? = null

    override var isFocussed: Boolean
        get() = item.state and Drawable.STATE_FOCUSSED != 0
        set(focussed) = if (focussed) {
            item.state = item.state or Drawable.STATE_FOCUSSED
        } else {
            item.state = item.state and Drawable.STATE_FOCUSSED.inv()
        }

    override var isSelected: Boolean
        get() = item.state and Drawable.STATE_SELECTED != 0
        set(selected) = if (selected) {
            item.state = item.state or Drawable.STATE_SELECTED
        } else {
            item.state = item.state and Drawable.STATE_SELECTED.inv()
        }

    override var isTouched: Boolean
        get() = item.state and Drawable.STATE_TOUCHED != 0
        set(touched) = if (touched) {
            item.state = item.state or Drawable.STATE_TOUCHED
        } else {
            item.state = item.state and Drawable.STATE_TOUCHED.inv()
        }

    override var isActive: Boolean
        get() = item.state and Drawable.STATE_ACTIVE != 0
        set(active) = if (active) {
            item.state = item.state or Drawable.STATE_ACTIVE
        } else {
            item.state = item.state and Drawable.STATE_ACTIVE.inv()
        }

    override fun getBounds(rect: RectF) {
        val bounds = item.bounds
        rect.set(bounds.leftf, bounds.topf, bounds.rightf, bounds.bottomf)
    }


    /**
     * Craw this drawable onto an android canvas. The canvas has an ofset
     * preapplied so the top left of the drawing is 0,0. This method itself,
     * just creates the canvas and passes the work to [.onDraw]
     */
    override fun draw(canvas: Canvas, theme: Theme<AndroidStrategy, AndroidPen, AndroidPath>, scale: Double) {
        if (androidCanvas == null) {
            androidCanvas = AndroidCanvas(canvas, theme)
        } else {
            androidCanvas!!.setCanvas(canvas)
        }
        onDraw(androidCanvas!!.scale(scale), null)
    }

    protected open fun onDraw(androidCanvas: IAndroidCanvas, clipBounds: Rectangle?) {
        item.draw(androidCanvas, clipBounds)
    }
}
